gssapi/WrapToken.go

```diff
@@ -0,0 +1,230 @@
+	}
+	return encType.GetChecksumHash(key.KeyValue, checksumMe, keyUsage)
+}
+
+// Build a header suitable for a checksum computation
+func getChecksumHeader(flags byte, senderSeqNum uint64) []byte {
+	header := make([]byte, 16)
+	copy(header[0:], []byte{0x05, 0x04, flags, 0xFF, 0x00, 0x00, 0x00, 0x00})
+	ENC.PutUint64(header[8:], senderSeqNum)
+	return header
+}
+
+// Compute the payload + header checksum with the provided key and usage,
+// and compare it to the checksum present in the token
+func (wt *WrapToken) VerifyCheckSum(key types.EncryptionKey, keyUsage uint32) (bool, error) {
+	computed, cErr := wt.ComputeCheckSum(key, keyUsage)
+	if cErr != nil {
+		return false, cErr
+	}
+	if !bytes.Equal(computed, wt.CheckSum) {
+		return false, errors.New(
+			fmt.Sprintf("Checksum mismatch. Computed: %s, Contained in token: %s",
+				hex.EncodeToString(computed), hex.EncodeToString(wt.CheckSum)))
+	}
+	return true, nil
+}
+
+// Parse a wrap token
+// if expectFromAcceptor is true, we expect the token to have been emitted by the gss acceptor,
+// and will check the according flag
+func UnmarshalWrapToken(b []byte, expectFromAcceptor bool) (*WrapToken, error) {
+	// Check if we can read a whole header
+	if len(b) < 16 {
+		return nil, errors.New("bytes shorter than header length.")
+	}
+	// Is the Token ID correct?
+	if !bytes.Equal(GSSWrapTokenID[:], b[0:2]) {
+		return nil, errors.New(
+			fmt.Sprintf("Wrong Token ID. Expected %s, was %s",
+				hex.EncodeToString(GSSWrapTokenID[:]),
+				hex.EncodeToString(b[0:2])))
+	}
+	// Check the acceptor flag
+	flags := b[2]
+	isFromAcceptor := flags&0x01 == 1
+	if isFromAcceptor && !expectFromAcceptor {
+		return nil, errors.New("Unexpected acceptor flag is set. not expecting a token from the acceptor.")
+	}
+	if !isFromAcceptor && expectFromAcceptor {
+		return nil, errors.New("Expected acceptor flag is not set. expecting a token from the acceptor, not the initiator.")
+	}
+	// Check the filler byte
+	if b[3] != FillerByte {
+		return nil, errors.New(
+			fmt.Sprintf("Unexpected filler byte: expecting 0xFF, was %s ", hex.EncodeToString(b[3:4])))
+	}
+	checksumL := ENC.Uint16(b[4:6])
+	// Sanity check on the checksum length
+	if int(checksumL) > len(b)-HdrLen {
+		return nil, errors.New(
+			fmt.Sprintf("Inconsistent checksum length. %d bytes to parse, checksum length is %d", len(b), checksumL))
+	}
+	rrc := ENC.Uint16(b[6:8])
+	seqNum := ENC.Uint64(b[8:16])
+	payload := b[16 : len(b)-int(checksumL)]
+	checksum := b[len(b)-int(checksumL):]
+	return &WrapToken{
+		Flags:    flags,
+		EC:       checksumL,
+		RRC:      rrc,
+		SND_SEQ:  seqNum,
+		Payload:  payload,
+		CheckSum: checksum,
+	}, nil
+}
+
+// Build a new initiator token (acceptor flag will be set to 0) and compute the authenticated checksum.
+// Other flags are set to 0, and the RRC and sequence number are initialized to 0.
+// Note that in certain circumstances you may need to provide a sequence number that has been defined earlier,
```

Comment:
Do you know where this sequence number would come from? Just interested to understand what would be involved to add support for this.

Reply:
From my little understanding, this is the sequence number that is set in the Authenticator, when it is sent as part of the initial negotiation request. From the RFC, Wrap or MIC tokens that accompany each message of an exchange when the "integrity" or "confidentiality" mode is chosen must increment the numbers to prevent replay attacks. 

When these tokens are part of an authentication handshake, however, at least the server I'm testing against (hadoop 3.0.0 namenode) does not care. If I switched to per-message encryption/signing, it might well start to check this, though I'm not planing to test this for now.

Reply:
So, to add support for this: the caller could keep the authenticator around and fill in the seq num by himself.
Or, even better: we could add a func to the Authenticator that builds a token with the correct seqnum and signs it using the correct session key.